### ISSUE TITLE
Develop & broadcast best coding practices

### ISSUE THREAD

**Issue by davidperrenoud** [fritzing/fritzing-app#1123]
_From [bryant.m...@gmail.com](https://code.google.com/u/100278494353872040962/) on June 15, 2010 19:29:15_

Most big projects follow some form of coding standards to synchronize development from multiple authors. An example is the Apache Geronimo project ( http://geronimo.apache.org/coding-standards.html ).

Considering the variety of methods people have coded this project I would think having some coding standards to follow would be awesome. I have seen a lot of code that has weird indentation because a single standard isn't being followed.

_Original issue: http://code.google.com/p/fritzing/issues/detail?id=1123_

**Comment by failiz** [comment]
I am not sure if the [wiki page](https://github.com/fritzing/fritzing-app/wiki/2.4-Coding-style) is enough to close this issue.

**Comment by failiz** [comment]
Ping

**Comment by KjellMorgenstern** [comment]
The wiki page refers to the qt codeing style, then add as few Fritzing specific rules as possible.
I could try to export the QtCreator specific configuration, do you think that would help?

**Comment by failiz** [comment]
Yes, I think that could be an easy way to start. Maybe also for Visual Studio if you have it configured. Thus, we could use the auto-format feature of the IDE. However, there are maybe other things to address too. For example, different levels of logging/when to thrown exceptions, how to document the code, how to organize the imports, etc.

**Comment by KjellMorgenstern** [comment]
Well, one general rule would be to never go a directory up (` import "../some.h"` )  For exceptions, as a rule of thumb, don't throw any exceptions in qt: https://doc.qt.io/qt-5/exceptionsafety.html  . The guidelines are violated in the Fritzing code, so there would be huge refactoring to really establish them. Doxygen documentation is welcome, but please limit it to code you wrote, or really understand in detail.

**Comment by failiz** [comment]
never go a directory up ( import "../some.h" ) ? The code is full of these ones!
I was not aware of the exceptions, actually I think the simulator code throws an exception in some cases. I will change it. Even if we do not refactor the code, I think we should try to avoid them in new code. 
I think the best would be if you update the wiki page with all this info, so new developers are aware of these things.
